File: clients/kotlin/src/main/kotlin/model/PlayerMessageGame.kt
package model

import util.StreamUtil

abstract class PlayerMessageGame {
    @Throws(java.io.IOException::class)
    abstract fun writeTo(stream: java.io.OutputStream)
    companion object {
        @Throws(java.io.IOException::class)
        fun readFrom(stream: java.io.InputStream): PlayerMessageGame {
            when (StreamUtil.readInt(stream)) {
                CustomDataMessage.TAG -> return CustomDataMessage.readFrom(stream)
                ActionMessage.TAG -> return ActionMessage.readFrom(stream)
                else -> throw java.io.IOException("Unexpected discriminant value")
            }
        }
    }

    class CustomDataMessage : PlayerMessageGame {
        lateinit var data: model.CustomData
        constructor() {}
        constructor(data: model.CustomData) {
            this.data = data
        }
        companion object {
            val TAG = 0
            @Throws(java.io.IOException::class)
            fun readFrom(stream: java.io.InputStream): CustomDataMessage {
                val result = CustomDataMessage()
                result.data = model.CustomData.readFrom(stream)
                return result
            }
        }
        @Throws(java.io.IOException::class)
        override fun writeTo(stream: java.io.OutputStream) {
            StreamUtil.writeInt(stream, TAG)
            data.writeTo(stream)
        }
    }

    class ActionMessage : PlayerMessageGame {
        lateinit var action: model.Versioned
        constructor() {}
        constructor(action: model.Versioned) {
            this.action = action
        }
        companion object {
            val TAG = 1
            @Throws(java.io.IOException::class)
            fun readFrom(stream: java.io.InputStream): ActionMessage {
                val result = ActionMessage()
                result.action = model.Versioned.readFrom(stream)
                return result
            }
        }
        @Throws(java.io.IOException::class)
        override fun writeTo(stream: java.io.OutputStream) {
            StreamUtil.writeInt(stream, TAG)
            action.writeTo(stream)
        }
    }
}

File: clients/javascript/model/player-message-game.js
const CustomData = require('./custom-data').CustomData;
const Versioned = require('./versioned').Versioned;
class PlayerMessageGame {
    static async readFrom(stream) {
        let discriminant = await stream.readInt();
        if (discriminant == CustomDataMessage.TAG) {
            return await CustomDataMessage.readFrom(stream);
        }
        if (discriminant == ActionMessage.TAG) {
            return await ActionMessage.readFrom(stream);
        }
        throw new Error("Unexpected discriminant value");
    }
}

class CustomDataMessage extends PlayerMessageGame {
    constructor(data) {
        super();
        this.data = data;
    }
    static async readFrom(stream) {
        let data;
        data = await CustomData.readFrom(stream);
        return new CustomDataMessage(data);
    }
    async writeTo(stream) {
        await stream.writeInt(CustomDataMessage.TAG);
        let data = this.data;
        await data.writeTo(stream);
    }
}
PlayerMessageGame.CustomDataMessage = CustomDataMessage;
CustomDataMessage.TAG = 0;
class ActionMessage extends PlayerMessageGame {
    constructor(action) {
        super();
        this.action = action;
    }
    static async readFrom(stream) {
        let action;
        action = await Versioned.readFrom(stream);
        return new ActionMessage(action);
    }
    async writeTo(stream) {
        await stream.writeInt(ActionMessage.TAG);
        let action = this.action;
        await action.writeTo(stream);
    }
}
PlayerMessageGame.ActionMessage = ActionMessage;
ActionMessage.TAG = 1;
module.exports = { PlayerMessageGame: PlayerMessageGame }
